File: frontend/lib/tax.js
export const DEFAULT_TAX_RATE = 0.19
export const CA_TAX_RATE = 0.0725

export function getItemTaxRate(item) {
  if (item?.tasa_impuesto != null) return Number(item.tasa_impuesto)
  return DEFAULT_TAX_RATE
}

/** Impuesto por línea con descuento repartido proporcionalmente. */
export function calculateCartTax(items, subtotal, discount = 0) {
  if (!items?.length || subtotal <= 0) return 0

  const taxableSubtotal = Math.max(subtotal - discount, 0)
  let totalTax = 0

  for (const item of items) {
    const lineSubtotal = Number(item.subtotal ?? item.precio_unitario * item.cantidad)
    if (lineSubtotal <= 0) continue
    const share = lineSubtotal / subtotal
    const lineTaxable = taxableSubtotal * share
    totalTax += lineTaxable * getItemTaxRate(item)
  }

  return Math.round(totalTax)
}

export function formatTaxLabel(items) {
  const rates = [...new Set((items || []).map(getItemTaxRate))]
  if (rates.length === 1) {
    const pct = rates[0] * 100
    const formatted = Number.isInteger(pct) ? String(pct) : pct.toFixed(2).replace(/\.?0+$/, '')
    return `Tax (${formatted}%)`
  }
  return 'Tax'
}

File: backend/app/utils/tax.py
"""
Impuestos por producto — tasa en producto.contenido_extra.tax_rate (decimal).
Sin override: 19% (tasa global del catálogo).
"""
from __future__ import annotations

from typing import Any, Iterable, Optional, Tuple

DEFAULT_TAX_RATE = 0.19
CA_TAX_RATE = 0.0725


def get_tax_rate_from_producto(producto: Any) -> float:
    """Lee tasa desde contenido_extra.tax_rate o devuelve DEFAULT_TAX_RATE."""
    if not producto:
        return DEFAULT_TAX_RATE
    extra = getattr(producto, "contenido_extra", None) or {}
    if not isinstance(extra, dict):
        return DEFAULT_TAX_RATE
    rate = extra.get("tax_rate")
    if rate is None:
        return DEFAULT_TAX_RATE
    try:
        return float(rate)
    except (TypeError, ValueError):
        return DEFAULT_TAX_RATE


def calcular_impuestos(
    lineas: Iterable[Tuple[float, float]],
    subtotal: float,
    descuento: float = 0,
) -> float:
    """
    Reparte el descuento proporcionalmente y aplica la tasa de cada línea.
    lineas: (subtotal_línea, tasa_impuesto)
    """
    lineas_list = [(float(s), float(r)) for s, r in lineas if s > 0]
    if not lineas_list or subtotal <= 0:
        return 0.0

    taxable_subtotal = max(subtotal - descuento, 0.0)
    total_tax = 0.0
    for line_subtotal, rate in lineas_list:
        share = line_subtotal / subtotal
        line_taxable = taxable_subtotal * share
        total_tax += line_taxable * rate

    return round(total_tax, 2)
